### PR TITLE
Smart listener support for better scalability

### PR DESCRIPTION
**Pull request by asimarslan**

### CODE REVIEW

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 1]
```diff
@@ -14,113 +14,394 @@
...
+                    connection);
+
+            IClientMessage clientMessage;
+            try
+            {
+                clientMessage = ThreadUtil.GetResult(future);
             }
             catch (Exception e)
             {
                 throw ExceptionUtil.Rethrow(e);
```
Comment (ihsandemir) [inline Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: is this rethrow needed?

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 2]
```diff
@@ -14,113 +14,394 @@
...
             }
+
+            var serverRegistrationId = registrationKey.ResponseDecoder(clientMessage);
+            var correlationId = registrationKey.RegistrationRequest.GetCorrelationId();
+            var registration = new EventRegistration(serverRegistrationId, correlationId, connection);
+
+            Debug.Assert(registrationMap != null, "registrationMap should be created!");
+            registrationMap.TryAdd(connection, registration);
```
Comment (ihsandemir) [inline Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: TryAdd or Just Add?

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 3]
```diff
@@ -14,113 +14,394 @@
...
+            Debug.Assert(!Thread.CurrentThread.Name.Contains("eventRegistration"));
+
+            ConcurrentDictionary<ClientConnection, EventRegistration> registrationMap;
+            if (_registrations.TryGetValue(registrationKey, out registrationMap) &&
+                registrationMap.ContainsKey(connection))
+            {
+                return;
+            }
+            var future = ((ClientInvocationService) _client.GetInvocationService())
+                .InvokeListenerOnConnection(registrationKey.RegistrationRequest, registrationKey.EventHandler,
```
Comment (ihsandemir) [inline Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: Java does invokeUrgent here. I did not see urgent invocation at CSharp, does it exist?

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 4]
```diff
@@ -14,113 +14,394 @@
...
+                    try
+                    {
+                        _connectionManager.GetOrConnect(member.GetAddress());
+                        return;
+                    }
+                    catch (Exception)
+                    {
+                        // ignored as this exception is retryed
+                    }
+                } while (_client.GetLifecycleService().IsRunning());
```
Comment (ihsandemir) [inline Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: Differs from Java: Java sleeps invocationRetryPauseMillis between each round of tries.

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 5]
```diff
@@ -14,113 +14,394 @@
...
+            }, token)).ToArray();
+
+            var timeoutMillis = ((ClientInvocationService) _client.GetInvocationService()).InvocationTimeoutMillis;
+            if (!Task.WaitAll(tasks, timeoutMillis))
+            {
+                throw new TimeoutException("Registering listeners is timed out.");
```
Comment (ihsandemir) [inline Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: If task timeouts, then is there a way to force kill the task?

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 6]
```diff
@@ -14,113 +14,394 @@
...
+
+            var eventTreadCount = EnvironmentUtil.ReadEnvironmentVar("hazelcast.client.event.thread.count") ??
+                                  DefaultEventThreadCount;
+
+            var eventQueueCapacity = EnvironmentUtil.ReadEnvironmentVar("hazelcast.client.event.queue.capacity") ??
+                                     DefaultEventQueueCapacity;
+
+            _eventExecutor = new StripedTaskScheduler(eventTreadCount, eventQueueCapacity, client.GetName() + ".event");
+            _registrationScheduler =
+                new StripedTaskScheduler(1, eventQueueCapacity, client.GetName() + ".eventRegistration");
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: There is no builtin singleThreaded Task executor so I used it as it is already tested code. No need to add a separate custom executor

File: Hazelcast.Net/Hazelcast.Util/StripedTaskScheduler.cs [hunk 7]
```diff
@@ -29,22 +29,27 @@ internal class StripedTaskScheduler : TaskScheduler, IDisposable
...
-        public StripedTaskScheduler(int numberOfThreads)
+        public StripedTaskScheduler(int numberOfThreads, int maximumQueueCapacity=1000000, string threadNamePrefix="hz-striped-scheduler")
         {
             if (numberOfThreads < 1) throw new ArgumentOutOfRangeException("numberOfThreads");
 
             _numberOfThreads = numberOfThreads;
+            // `maximumQueueCapacity` is the given max capacity for this executor. Each worker in this executor should consume
+            // only a portion of that capacity. Otherwise we will have `threadCount * maximumQueueCapacity` instead of
+            // `maximumQueueCapacity`.
+            var perThreadMaxQueueCapacity = (int) Math.Ceiling(1D * maximumQueueCapacity / numberOfThreads);
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Util/StripedTaskScheduler.cs]: I changed this part to make it same with Java side. Please see Java implementation.

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs (hunk 1 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: yes outer level try-catch block wait the inner-exceptions not a wrapper of AggregateException, we need to peel it for them

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs (hunk 3 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: It's backpressure related call. C# doesn't have that yet.

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 8]
```diff
@@ -14,113 +14,394 @@
...
+        private readonly ClientConnectionManager _connectionManager;
+
+        private Timer _connectionReopener;
+
+        private readonly ConcurrentDictionary<ClientConnection, ICollection<ListenerRegistrationKey>>
+            _failedRegistrations;
 
-        private readonly ConcurrentDictionary<string, long> _registrationMap = new ConcurrentDictionary<string, long>();
+        private readonly
+            ConcurrentDictionary<ListenerRegistrationKey, ConcurrentDictionary<ClientConnection, EventRegistration>>
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: done

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs [hunk 9]
```diff
@@ -14,113 +14,394 @@
...
-                    task = _client.GetInvocationService()
-                        .InvokeListenerOnRandomTarget(request, handler, responseDecoder);
-                }
-                else
+                var userRegistrationId = Guid.NewGuid().ToString();
+
+                var registrationKey =
+                    new ListenerRegistrationKey(userRegistrationId, registrationMessage, responseDecoder, eventHandler);
+
+                _registrations.TryAdd(registrationKey, new ConcurrentDictionary<ClientConnection, EventRegistration>());
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: return value means there is a value already exist or not so we can ignore it as we just add a new value if not exist one.

[the remark this answers is not in the document]
Comment (ihsandemir) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: Yes, exactly. I am not sure if the same key will happen or not not but Java behaves by replacing it. (.Net seems to have different methods like AddOrUpdate to achieve this). I am ok with this, just wanted to raise the awareness for the difference.

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs (hunk 5 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: cancellation added

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs (hunk 2 above)
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: replaced with Add functionality (index operator [] )

File: Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs]: As new uuid is created, TryAdd and Add does not matter, it will always be empty

**Pull request merged**